expanded weights: conv faster rule

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70141
* #73696
* #73694
* #73693
* __->__ #73692

Differential Revision: [D34719302](https://our.internmc.facebook.com/intern/diff/D34719302)

Commit:
expanded weights: conv faster rule

[ghstack-poisoned]

zou3519 commented on test/test_expanded_weights.py:
```diff
@@ -198,6 +200,47 @@ def test_expanded_weight_error(self, device):
+
+        expected = [torch.stack(grad) for grad in zip(*expected)]
+        for (res, exp) in zip(result, expected):
+            assert torch.allclose(res, exp, atol=1e-4, rtol=5e-5)
```
nit: prefer self.assertEqual

zou3519 commented on torch/nn/utils/_expanded_weights/conv_expanded_weights.py:
```diff
@@ -0,0 +1,67 @@
+from .expanded_weights_impl import implements_per_sample_grads
+from .expanded_weights_utils import forward_helper
+
+THRESHOLD = 257
```
Is this used at all?

zou3519 commented on torch/nn/utils/_expanded_weights/conv_expanded_weights.py:
```diff
@@ -0,0 +1,67 @@
+
+THRESHOLD = 257
+
+@implements_per_sample_grads(F.conv1d)
+class Conv1dPerSampleGrad(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, kwarg_names, *expanded_args_and_kwargs):
+        if any([isinstance(i, str) for i in expanded_args_and_kwargs]):
+            raise RuntimeError("Expanded Weights does not support convolution padding as a string. "
+                               "Please file an issue to prioritize support")
+        expanded_args, expanded_kwargs = conv_args_and_kwargs(kwarg_names, expanded_args_and_kwargs)
+        output = forward_helper(F.conv1d, expanded_args, expanded_kwargs)
+        ctx.input, ctx.weight = expanded_args
+        ctx.bias = expanded_kwargs['bias']
+        ctx.stride, ctx.padding = expanded_kwargs['stride'], expanded_kwargs['padding']
+        ctx.dilation, ctx.groups = expanded_kwargs['dilation'], expanded_kwargs['groups']
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return conv_backward(F.conv1d, ctx, grad_output)
+
```
Can we de-duplicate this more?

```
class ConvPerSampleGrad(torch.autograd.Function):
    @staticmethod
    def forward(ctx, kwarg_names, *expanded_args_and_kwargs):
        if any([isinstance(i, str) for i in expanded_args_and_kwargs]):
            raise RuntimeError("Expanded Weights does not support convolution padding as a string. "
                               "Please file an issue to prioritize support")
        expanded_args, expanded_kwargs = conv_args_and_kwargs(kwarg_names, expanded_args_and_kwargs)
        conv_fn = get_conv_fn(expanded_args) # selects 1d, 2d, or 3d.
        output = forward_helper(conv_fn, expanded_args, expanded_kwargs)
        ctx.input, ctx.weight = expanded_args
        ctx.bias = expanded_kwargs['bias']
        ctx.stride, ctx.padding = expanded_kwargs['stride'], expanded_kwargs['padding']
        ctx.dilation, ctx.groups = expanded_kwargs['dilation'], expanded_kwargs['groups']
        return output

    @staticmethod
    def backward(ctx, grad_output):
        return conv_backward(conv_fn, ctx, grad_output)

# IDK if the decorator actually works like this
implements_per_sample_grads(F.conv1d)(ConvPerSampleGrad)
implements_per_sample_grads(F.conv2d)(ConvPerSampleGrad)
implements_per_sample_grads(F.conv3d)(ConvPerSampleGrad)
```

samdow replied:
Fixed--thankfully decorators what we would hope in that situation

zou3519 commented on torch/nn/utils/_expanded_weights/conv_expanded_weights.py:
```diff
@@ -0,0 +1,67 @@
+                               "Please file an issue to prioritize support")
+        expanded_args, expanded_kwargs = conv_args_and_kwargs(kwarg_names, expanded_args_and_kwargs)
+        output = forward_helper(F.conv2d, expanded_args, expanded_kwargs)
+        ctx.input, ctx.weight = expanded_args
+        ctx.bias = expanded_kwargs['bias']
```
You don't need to save bias for backward -- you only need to save the shape of bias, right?

For ctx.input and ctx.weight -- we don't need to save input unless weight requires_grad. We don't need to save weight unless input requires grad.

These changes will help improve our memory utilization (I don't remember if Opacus does these already, if not, then all the better -- we're making a better system :) )

samdow replied:
Sadly with the current design we do need to save bias for the backward pass. Because expanded weights are producing gradients whose shape adds a batch dimension, I'm not having us use autograd directly but instead set the grad_sample attribute directly during the backward pass. Because it's setting the attribute directly, we needed to save the parameter

zou3519 replied:
Gotcha. So you need to save the weight, do you need to save the input? The input doesn't have a grad_sample field to populate right?

samdow replied:
Good point! Updated. Looks like the remaining functions after this (embedding + norms) call a backwards function that takes in input so I didn't update those

zou3519 commented on torch/nn/utils/_expanded_weights/conv_utils.py:
```diff
@@ -0,0 +1,184 @@
+        return conv3dOpt
+
+conv_kwarg_defaults = {'bias': None, 'stride': 1, 'padding': 0, 'dilation': 1, 'groups': 1}
+def conv_args_and_kwargs(kwarg_names, expanded_args_and_kwargs):
+    args = expanded_args_and_kwargs[:len(expanded_args_and_kwargs) - len(kwarg_names)]
+    named_kwargs = expanded_args_and_kwargs[len(expanded_args_and_kwargs) - len(kwarg_names):]
+    unnamed_kwargs = args[2:]  # input and weight are the only two that can't be kwargs
+
+    conv_kwargs = tuple(conv_kwarg_defaults.keys())
+    kwargs = {key: value for (key, value) in zip(conv_kwargs[:len(unnamed_kwargs)] + kwarg_names, unnamed_kwargs + named_kwargs)}
+    kwargs = {key: kwargs.get(key, conv_kwarg_defaults[key]) for key in conv_kwarg_defaults}
+    return args[:2], kwargs
```
I'm a little confused about the packing scheme. The reason I'm confused is that we get here, we are assuming that input and weight aren't kwargs.

If the user does something like F.conv2d(input=input, weight=weight, ...), does the `__torch_function__` trigger with input and weight as kwargs, or does `__torch_function__` normalize everything so that input & weight are args and everything else is a kwarg?

That being said, if the only way we're saying that people can hit expanded weights is through the module API (e.g. nn.Conv1d) then this isn't really a problem

samdow replied:
Yeah...you're totally right that if conv2d is just directly passing the args + kwargs through like we think it is, this won't work as well. My plan/proposal for the future is to have __torch_function__ normalize the inputs because that's what the norm functions do but it's probably something that will need a little more discussion

zou3519 replied:
Okay, I see two potential paths forward here:

The first is to just leave a TODO about needing an argument parser for `__torch_function__` and filing an issue about it. I don't understand how other people use `__torch_function__` without an argument parser...

The second is try to work around the issue by writing our own argument normalizer. It's not too difficult to do this, the challenge (for the long-term) is keeping it in sync with the actual signature for conv2d. For example, if we had a `args, kwargs = normalize_conv_inputs(input, weight, bias, ...)` etc function that has the same exact signature (including default arguments!) as conv2d, then that could be a quick workaround

```
def normalize_conv_inputs(input, weight, bias, ...):
  args = (input, weight, bias)
  kwargs = ...
  return args, kwargs
```

samdow replied:
Great idea for the short term! Added the second idea

For how other people use it, I found some other old issues too that I can pull. My gut is that we should parse it so it's passed to `__torch_function__` in a standardized manner but there's some old issues (it seems like from before `__torch_dispatch__`) so I want to investigate those a little more before changing the codegen

zou3519 reviewed (approved):
LGTM, with comments though, please read.

- We should be able to deduplicate some logic so we don't have separate autograd.Function for the 3 convolution cases
- I'm a little worried that the unpacking/repacking scheme might not work depending on how `__torch_function__` decides to call things. For example, if a user does `F.conv2d(input=input, weight=weight, ...)`, does `__torch_function__` treat input and weight as positional arguments or kwargs?
- I didn't read the backward computation code in-depth but I assume you pulled them from Opacus and they are being tested

Commit:
Update on "expanded weights: conv faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: conv faster rule"




[ghstack-poisoned]

Commit:
Update on "expanded weights: conv faster rule"




[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34719302).

Commit:
Update on "expanded weights: conv faster rule"


Differential Revision: [D34719302](https://our.internmc.facebook.com/intern/diff/D34719302)

[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34719302).
